Add the ability to authenticate using only an API Token

syncrou opened the pull request.

syncrou commented:
@simon3z - Was hoping you could review this (tests coming) But It looks like this PR is duplicated effort - https://github.com/ManageIQ/manageiq-api-client-python/pull/22

syncrou commented:
@simon3z - Please review, this iteration includes tests.

simon3z commented on src/manageiq_client/api.py:
```diff
@@ -25,24 +25,34 @@ class ManageIQClient(object):
         self.response = None
         self._load_data()
 
+    def _build_auth(self, auth):
```
@syncrou one comment on the current code is that there are two code-paths (`len(auth) == 2` and `== 1`) that are way too similar (would need factoring out to a common function).

Also, let's remember that we probably want to support only a single (and possibly explicit) way of authenticating. The fact that we're not supporting both `dict` and `tuple`/`list` is probably for backward compatibility IIRC.

So all in all (to reduce the amount of repeated code and to focus on just a single auth format for new features) I'd suggest:

```python
    def _build_auth(self, auth):
      if isinstance(auth, dict):
        if "user" in auth:      
          self._auth = (auth["user"], auth["password"])
        if "token" in auth:
          headers['X-Auth-Token'] = auth['token']
          self._session.headers.update(headers)
      elif isinstance(auth, (tuple, list)):  # for backward compatibility
          self._auth = tuple(auth[:2])
      else:
          raise ValueError("Unknown values provider for auth")
```

cc @cben

cben replied:
agreed

syncrou replied:
@simon3z, @cben - Thanks for all the great feedback!  I'll get going on your suggested changes today.

simon3z commented:
@cben please review

cben commented on example.py:
```diff
@@ -4,8 +4,16 @@
 url = os.environ.get('MIQURL') or 'http://localhost:3000/api'
 username = os.environ.get('MIQUSERNAME') or 'admin'
 password = os.environ.get('MIQPASSWORD') or 'smartvm'
+token = { 'token': os.environ.get('MIQTOKEN') }
```
cosmetic: I'd do
```
token = os.environ.get('MIQTOKEN')
...
if token:
     ... MiqApi(url, dict(token=token))
```

cben commented on example.py:
```diff
@@ -4,8 +4,16 @@
+    client = MiqApi(url, token)
+else:
+    print("\nAuthenticating with the user credentials")
+    client = MiqApi(url, (username, password))
```
and here I'd  use  `dict(user=username, password=password)`
format for consistency

cben commented on src/manageiq_client/api.py:
```diff
@@ -25,24 +25,34 @@ class ManageIQClient(object):
         self._load_data()
 
+    def _build_auth(self, auth):
+        headers = {'Content-Type': 'application/json; charset=utf-8'}
```
this header belongs outside this function.
(and as currently written isnt even set in tuple code path)

cben commented on testing/test_api_credentials.py:
```diff
@@ -0,0 +1,27 @@
+def test_invalid_token_key():
+    with pytest.raises(KeyError):
+        MiqApi('http://www.test.com', {'Broken':'123234'})
+        MiqApi('http://www.test.com', {'Token':'123234'})
```
consider also testing correct inputs.

cben replied:
ah, you did, for token.  but not user,pw.

syncrou commented:
@simon3z, @cben - I believe I addressed all the feedback.  I found I had to add more conditional logic around `_build_auth` so I could raise an exception for empty auth credentials passed in as a `dict`. I found that issue [here](https://github.com/syncrou/manageiq-api-client-python/blob/c7f68cb3314bf67cf860007f2bf405d7ff7c4584/testing/test_api_credentials.py#L9).

simon3z commented:
@cben can you review / approve? Thanks.

cben commented on testing/test_api_credentials.py:
```diff
@@ -0,0 +1,35 @@
+
+
+def test_no_password_dict():
+    with pytest.raises(KeyError):
```
consider making this too a ValueError, for consistency (`if "user" in auth and "password" in auth`)
LGTM either way.

cben reviewed (approved):
:+1:

syncrou commented:
@cben - I made a small change to raise `ValueError` if both `auth['user']` and `auth['password']` aren't passed in via a `dict`.

Thanks for the suggestion.

cben commented:
LGTM :+1:

syncrou commented:
@simon3z - Do you see anything else that should be added / changed?

simon3z commented:
@syncrou just one thing, please squash the PR into a single commit and we'll merge. Thanks.

syncrou committed:
Add ability to authenticate using the api token

1. Allow for backwards compatibility if a username, password is passed in via a list
2. Only use the token for authentication if that is passed in
3. Add tests around passing authentication credentials

syncrou commented:
@simon3z - Commits are squashed as requested.
